### PR TITLE
Amendments

### PR DESCRIPTION
**Pull request by CatoTH**
Current state:
- Creating amendments: Selecting a paragraph and changing the text works. Saving works.
- Displaying the amendment (Single view)
- Displaying the amendment (embedded into base motion)
- Editing the amendment afterwards
- Evaluating the states / recommendations to decide if it's to be included in the final version of the motion.

Debatable design decisions:
- Displaying amendments in single view uses the motion-details view. This is just as with OpenSlides 2.
- Creating a amendment is a component of its own. This is different from the motion implementation, as we have a lot of amendment-specific stuff.
- Currently, only amending one paragraph is supported. However, the structures are prepared to handle multiple paragraphs in the future. Specifically, one amendment can multiple change objects. We had that in the database already, now we also have a view object for that (ViewMotionAmendedParagraph).

### CODE REVIEW

File: client/src/app/site/motions/services/diff.service.ts [hunk 1]
```diff
@@ -2006,4 +2053,50 @@ export class DiffService {
...
+
+    /**
+     * This is used to extract affected lines of a paragraph with the possibility to show the context (lines before
+     * and after) the changed lines and displaying the line numbers.
+     *
+     * @param paragraphNo
+     * @param origText - needs to be line-numbered
+     * @param newText
+     * @param lineLength
+     * @return DiffLinesInParagraph
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/diff.service.ts]: @return**s** also the parameters should have some descriptions

File: client/src/app/site/motions/services/diff.service.ts [hunk 2]
```diff
@@ -2006,4 +2053,50 @@ export class DiffService {
...
+                this.extractRangeByLineNumbers(diff, affected_lines.to, paragraph_line_range.to),
+                lineLength, affected_lines.to
+            );
+        }
+        const text = this.formatDiffWithLineNumbers(
+            this.extractRangeByLineNumbers(diff, affected_lines.from, affected_lines.to),
+            lineLength, affected_lines.from
+        );
+
+        return {
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/diff.service.ts]: this returns "some" object. I would advise to use the interface in the function or use `return {...} as DiffLinesInParagraph`

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 3]
```diff
@@ -161,6 +163,19 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
         await this.httpService.post(url, { nodes: motionIds });
     }
 
+    /**
+     * Returns an observable returning the amendments to a given motion
+     *
+     * @param motionId
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: requires @returns with some more documentation

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 4]
```diff
@@ -370,4 +385,79 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
 
         return diff;
     }
+
+    /**
+     * Given an amendments, this returns the motion affected by this amendments
+     *
+     * @param amendment
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: requires @returns

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 5]
```diff
@@ -370,4 +385,79 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
...
+     */
+    public getAmendmentBaseMotion(amendment: ViewMotion): ViewMotion {
+        return this.getViewModel(amendment.parent_id);
+    }
+
+    /**
+     * Splits a motion into paragraphs, optionally adding line numbers
+     *
+     * @param motion
+     * @param lineBreaks
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: requires @returns

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 6]
```diff
@@ -370,4 +385,79 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
...
+            const lineLength = motion.lineLength;
+            html = this.lineNumbering.insertLineNumbers(html, lineLength);
+        }
+        return this.lineNumbering.splitToParagraphs(html);
+    }
+
+    /**
+     * Returns all paragraphs that are affected by the given amendment in diff-format
+     *
+     * @param amendment
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: @returns

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 7]
```diff
@@ -370,4 +385,79 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
...
+            }
+            return this.diff.getAmendmentParagraphsLinesByMode(paraNo, baseParagraphs[paraNo], newText, lineLength);
+        }).filter((para: DiffLinesInParagraph) => para !== null);
+    }
+
+    /**
+     * Returns all paragraphs that are affected by the given amendment as unified change objects.
+     * In the paragraphs, only the
+     *
+     * @param amendment
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: @returns

File: client/src/app/site/motions/services/motion-repository.service.ts [hunk 8]
```diff
@@ -370,4 +385,79 @@ export class MotionRepositoryService extends BaseRepository<ViewMotion, Motion>
...
+    }
+
+    /**
+     * Returns all paragraphs that are affected by the given amendment in diff-format
+     *
+     * @param amendment
+     */
+    public getAmendedParagraphs(amendment: ViewMotion): DiffLinesInParagraph[] {
+        const motion = this.getAmendmentBaseMotion(amendment),
+            baseParagraphs = this.getTextParagraphs(motion, true),
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-repository.service.ts]: please use `const` all the time when declaring variables. Also in the other functions.

File: client/src/app/site/motions/components/amendment-create-wizard/amendment-create-wizard.component.html [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        <h2 translate>
+            Create amendment
+        </h2>
+    </div>
+</os-head-bar>
+
+<form [formGroup]='contentForm' (ngSubmit)='saveAmendment()'>
+    <mat-horizontal-stepper linear>
+        <mat-step [label]="'Select paragraph' | translate" [completed]="contentForm.value.selectedParagraph">
+            <ng-template matStepLabel translate>Select paragraph</ng-template>
```
[the remark this answers is not in the document]
Comment (CatoTH) [reply client/src/app/site/motions/components/amendment-create-wizard/amendment-create-wizard.component.html]: This seems to be a requirement by `mat-horizontal-stepper`: https://material.angular.io/components/stepper/overview

File: client/src/app/site/motions/components/motion-detail/motion-detail.component.ts [hunk 10]
```diff
@@ -211,14 +232,19 @@ export class MotionDetailComponent extends BaseViewComponent implements OnInit {
         } else {
             // load existing motion
             this.route.params.subscribe(params => {
-                this.repo.getViewModelObservable(params.id).subscribe(newViewMotion => {
+                const motionId: number = parseInt(params.id, 10);
```
[the remark this answers is not in the document]
Comment (CatoTH) [reply client/src/app/site/motions/components/motion-detail/motion-detail.component.ts]: ok, done. But what JS implementations are that?

**Comment by CatoTH** [comment]
Hint: a lot of the reformatting changes come from prettify-all

File: client/src/app/site/motions/components/motion-detail-diff/motion-detail-diff.component.html [hunk 11]
```diff
@@ -1,73 +1,77 @@
...
-                <span ng-if="change.original.getType(motion.getVersion(version).text) == 3">
-                  { change.other_description }
+            <a
+                href=""
+                (click)="scrollToChangeClicked(change, $event)"
+                [class.amendment]="isAmendment(change)"
+                [class.recommendation]="isChangeRecommendation(change)"
+            >
+                <span *ngIf="change.getLineFrom() >= change.getLineTo() - 1" class="line-number">
+                    {{ 'Line' | translate }} {{ change.getLineFrom()
```
Comment (tsiegleauq) [inline client/src/app/site/motions/components/motion-detail-diff/motion-detail-diff.component.html]: looks odd. Perhaps try to put that line into a `<span>` or a `<p>`?

try out [here ](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeArgGwHwB0qqYCWeBAzgA4CGUABAFRQDmAkgGYC8u4AFjU3AB0AmABkiUOADEAThAC2ACgCUtbB1pg+zISPGSAKhBW0AtLQCM3TZipkyXEMUmmo6eQCM4M7qWDBaAHJ9OEDaAB9aGBkaMlt4WgBfRNp-TW0BYTgxCWk5JVVk-FRKGgZmdkciMgBhDLgAJThIeXkEABMqGCJoRS1+OGVfVAB6UqhSUfHJkeIZrFJ8EAAaEAgKbugyZFAqGTkAdwAFPYRtlCoANwgidpWQDxiwAGtsgGVqMAkmZGj0OFWPBg8kwAHUeER4KUwHA3mdIURLpCAJ7IcD2e4SMjeGBHGJMeRUZBsKiYbGrABWZAAHgAhJ6vGBvKhtELE0nkkBU6lvb6YOAARXQEHg7LJAJA1Bk2JkaP4WD29woMgkMFBtxgPGQFgATAAGVbKiDY0ExCho5VwGWXOD3GRwACO6CI9rxVAJRKQJPFq2x8iIvxk-19fMFwtFXo5EpgVA86vamuQABZVtEqERnEwagpCWioNBbat0NiDLHzt7sckgA)

**Comment by emanuelschuetze** [comment]
I retested your pull request. I see three open todos which are blockers for me:
* [ ] error after creating amendment without changes
* [ ] we need a migration file (maybe @FinnStutzenstein could help you)
* [ ] the identifier of amendments should use configured prefix (e.g. "B01-Ä01").
We want to merge this pullrequest until friday to deliver new test version on friday. Can you update your pull request in the next days? Thx.

**Comment by CatoTH** [comment]
@emanuelschuetze 
I fixed the first bug (no changes).
Migrations for what exactly? The workflow states?
How do you configure these prefixes?

@tsiegleauq 
What "plus operator" did you refer to?
(For octal numbers, that should be prevented by explicitly specifying the 10 as numbering base, though)

**Comment by emanuelschuetze** [comment]
Retested:
* createing amendments without changes works now (it's possible to view and delete it, but there is still a error after creating: `TypeError: Cannot read property 'from' of null at motion-repository.service.ts:505`
* if you set the config value `motions_amendments_prefix` the identifier for amendment should use this prefix (e.g. B01-Ä01). The identifier is set by server. Any idea why it does not work anymore?
* db migration is still missiong: create new db with current master. Then checkout this PR. Restart server (`runserver`); server failed with: `django.db.utils.OperationalError: no such column: motions_state.merge_amendment_into_final` -> we have to add a new migration file for this.

**Comment by CatoTH** [comment]
@emanuelschuetze: Point 1 is done.
As for the prefix: as far as I can see it, this was done in the backend in v2, or not? I only could find references to motions_amendments_prefix in the python-scripts.

For migrations, I wrote something about it before. I can create a migration, however the problem is: how shall this new field being pre-filled? It needs to be 0, 1 or -1, but as the entries in `motion_state` are (potentially) arbitrary, I don't see a really reliable way of determining them. Except for hoping that nobody has changed the default values - in which case we can assume "Accepted" = 1 and "Rejected" = -1.

**Comment by emanuelschuetze** [comment]
> Point 1 is done.

Confirmed. Thanks!

> As for the prefix

I rested with new database. Now it works as expected - like in 2.3. Sorry...

>  I can create a migration, however the problem is: how shall this new field being pre-filled?

Suggestion: Check the [css_class](https://github.com/OpenSlides/OpenSlides/blob/master/openslides/motions/models.py#L1061) and set +1 for 'success' (green), -1 for 'danger' (red) and 0 for the rest.
The quick-and-dirty-migration would be: set all to 0 and the admin has to update the values in the workflow manage view.

**Comment by tsiegleauq** [comment]
rebased on current master, tested again and meld into one.

**Comment by normanjaeckel** [comment]
This pull request has a breaking change. Please correct or revert.

There is a new field for states "merge_amendment_into_final" with small integers. In the default workflows we use a boolean field which can not work. Also the meaning of this new flag is not clear. If a motion is in this state, the amendment should be merged into final version? This looks like a boolean flag and not a integer.

**Comment by CatoTH** [comment]
@normanjaeckel what exactly is broken?
As for why a boolean field is not enough, I think my comment here should clarify it. There needs to be an "undefined" state: https://github.com/OpenSlides/OpenSlides/pull/3975/commits/afad235bad39333a038e6c48b1af07dc249e437c#diff-0222f0cfc8faa48bb63a018ea216248eR1104

**Comment by normanjaeckel** [comment]
In openslides/motions/signals.py there is now a flag with a boolean value: https://github.com/OpenSlides/OpenSlides/pull/3975/files#diff-4a75caf0e252ecff50d71d40a27f20fdR28

**Comment by CatoTH** [comment]
@normanjaeckel you're right. Here's the bugfix-pr: https://github.com/OpenSlides/OpenSlides/pull/4043

**Comment by normanjaeckel** [comment]
The hint-text is not self explaining. What does it mean "status is 1"? A motion starts at first state of a workflow and then switches from state to state. Final state is something like "accepted" or "not decided". For states like "accepted" we may add an extra boolean flag that the machine knows that this motion (which is an amendmend) should be merged into its "parent", right? In first state and also in states like "rejected" the flag remains False, right?

**Comment by normanjaeckel** [comment]
Ok I understand now:

state not decided should be -1 also.

**Comment by CatoTH** [comment]
added an "with an merge_amendment_into_final-value of" :)

**Pull request merged**